Add type checking to Transform for Point/Point3D

mcdaniel67 opened the pull request:
I added type checking to transform for Point and Point3D. Now when we call transform it checks that the function argument is a matrix of the correct dimension. I also added tests for the type checking.

mcdaniel67 committed:
Added type checking to Transform

I added type checking to the transform code for Point and Point3D. I also added new tests to ensure that the changed code functions properly. Now before
a transformation is applied to the point we check that the supplied variable is a matrix and that it is of the correct dimension.

mcdaniel67 committed:
Readded tests I removed

I removed some tests when I was working on this originally. Those have been added back. I also removed the Matrix Transpose import from point.py.

debugger22 commented:
@akshayah3 Do you want to review this one?

toolforger commented on sympy/geometry/point.py:
```diff
@@ -516,8 +516,12 @@ def transform(self, matrix):
+            return Point(*(Matrix(1, 3, [x, y, 1])*matrix).tolist()[0][:2])
+        else:
+            raise ValueError("The argument to the transform function must be " \
+            + "a 3x3 matrix")
```
It's better to use this pattern:

```
if not isinstance(...) or not matrix.shape == ...:
  raise ValueError(...)
# original code goes here
```

This unnests code, so it's easier to read. (I.e. the productive code (original lines 519 f.) isn't buried deep down in a nested `if`.)
Also, if somebody wants to add additional checks, or split the check in two, adding an additional check doesn't mean you have to touch the productive code.
It also doesn't touch the original code, that makes it easier to see who authored it in `git annotate`.

toolforger commented on sympy/geometry/point3d.py:
```diff
@@ -513,10 +513,13 @@ def transform(self, matrix):
-        x, y, z = self.args
-        m = Transpose(matrix)
-        return Point3D(*(Matrix(1, 4, [x, y, z, 1])*m).tolist()[0][:3])
+        if isinstance(matrix, Matrix) and matrix.shape == (4, 4):
```
Same remark applies here.
Also, the check "is this a square matrix?" looks like it's generally useful. Maybe add an `is_square_matrix(size)` function.
I'm not sure that type checking in this way is a good idea. Python's approach is "duck typing", i.e. you simply call the appropriate functions and this will already do the crashing for you. So the `isinstance` check can go, it's enough to ask for the `shape` and let it crash if it's not a matrix (nor something that has a shape).
If an `isinstance` check is really desired, you should check for `MatrixBase`, because this function is supposed to work for `ImmutableMatrix` as well as for `Matrix` (I think).

mcdaniel67 commented:
That absolutely makes sense, I will switch the code so that the size check comes first. It also makes sense to check against MatrixBase now that I've looked at the ImmutableMatrix class. 

I can also add the is_square_matrix function, would this function just go into the MatrixBase class? Also, does it make more sense to pass a size into this function, or to just compare the col/row size and return true only if they match?

toolforger commented:
I guess yes it should go into `MatrixBase`. It's somewhat trivial, but ah well.
I guess `is_square` would be enough... hmm, well, no, `is_square_matrix` is fine, in a duck-typing regime, function names need to be reasonably unique across classes. Still, stick with the conventions that you find in `MatrixBase`. (I'd actually expect such a function to already exist. If it doesn't, it's a bit of an omission, squareness is an important property of matrices.)

mcdaniel67 committed:
Fixed the ordering

I switched the ordering/manner of the type checking. Now, the ValueError is before the code for the transform functions. I also switched away from using is_instance.

mcdaniel67 commented:
There is an `is_square` property for `MatrixBase`, so I didn't add the function. If we think it would still be useful as a standalone function I would be more than happy to add it, it wouldn't take much time.

toolforger commented:
Am 09.06.2015 um 21:00 schrieb Kyle McDaniel:
> There is an `is_square` property for `MatrixBase`, so I didn't add the function. If we think it would still be useful as a standalone function I would be more than happy to add it, it wouldn't take much time.

No, just call `is_square` on the object.
If it's not a matrix, it won't have `is_square` and people will see the 
normal exception that happens for type errors (it's not an easily 
understandable exception, but it's a general problem that we can't solve 
here).

certik commented on sympy/geometry/point.py:
```diff
@@ -516,9 +516,19 @@ def transform(self, matrix):
+            col, row = matrix.shape
+            valid_matrix = matrix.is_square and col == 3
+        except AttributeError:
+            # We hit this block if matrix argument is not actually a Matrx.
```
Matrx -> Matrix

certik commented on sympy/geometry/point3d.py:
```diff
@@ -513,6 +513,15 @@ def transform(self, matrix):
+            col, row = matrix.shape
+            valid_matrix = matrix.is_square and col == 4
+        except AttributeError:
+            # We hit this block if matrix argument is not actually a Matrx.
```
Matrx -> Matrix

certik commented:
This looks good to me, just a minor typo.

mcdaniel67 committed:
Fixed a typo

There was a typo in a comment, which I fixed.

mcdaniel67 commented:
I fixed the typo. I can't believe I didn't catch that!

mcdaniel67 commented:
The Travis tests are timing out. The two different commits timed out on two different tests and one test that was the same for both of them `Python 3.4 TEST_SLOW = "true" SPLIT="2/2"` was the one that was the same for both.

certik commented:
That's fine, we can merge it.

However, can you please resolve conflicts with master? I think some of the other PRs that we merged made some conflicts.

mcdaniel67 committed:
Merge upstream with my fork

Conflicts:
	sympy/geometry/point.py
	sympy/geometry/point3d.py (removed)
I merged point.py by accepting all the changes made in the upstream master branch, then splicing in my changes to transform for Point2D/Point3D.
point3d.py was removed on master of sympy/sympy, so I removed mine to match as well. This /should/ resolve the conflicts.

mcdaniel67 commented:
I merged to hopefully resolve the conflicts with master. I'm not entirely sure that I did it correctly, does the latest commit look correct?

akshayah3 commented:
Looks good. +1 from my side.

certik commented:
+1 from me as well if tests pass.

certik commented:
The timeouts are unrelated.
